### PR TITLE
Rdm 11710

### PR DESCRIPTION
**Pull request by KleoG**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-11710

### Change description ###
Update API Client Library to include "PATCH" operation
Update bintray to jitpack

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java:11]: please add @JsonIgnoreProperties(ignoreUnknown = true)

File: src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLRequest.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.hmcts.reform.ccd.document.am.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+
+@NoArgsConstructor
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLRequest.java:9]: try removing @NoArgsConstructor

File: src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package uk.gov.hmcts.reform.ccd.document.am.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+
+@NoArgsConstructor
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java:9]: try removing @NoArgsConstructor

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    private static final boolean PERMANENT = false;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Autowired
+    private CaseDocumentClientApi caseDocumentClientApi;
+
+    @BeforeAll
+    static void beforeAll(){
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:68]: remove please

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+    @BeforeEach
+    void setUp() {
+        objectMapper.registerModule(new JavaTimeModule());
+        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+
+
+    }
+
+    @AfterAll
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:79]: remove if not required

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        System.out.println(finalResponse.getBody());
+
+        assertEquals(HttpStatus.OK, finalResponse.getStatusCode());
+    }
+
+    @Test
+    void getDocumentBinaryTest() throws IOException {
+        ResponseEntity response = new ResponseEntity(HttpStatus.OK);
+
+        stubForDocumentBinary(response);
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:130]: you need to match request params / body in the wire-mock stub

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+        stubForDocumentMetaData(response);
+
+        Document finalResponse = caseDocumentClientApi.getMetadataForDocument(
+            TOKEN,
+            BEARER + TOKEN,
+            DOCUMENT_ID
+        );
+
+        System.out.println(response.createdOn);
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:156]: remove sou

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+        ResponseEntity response = new ResponseEntity(HttpStatus.OK);
+
+        stubForUpload(response);
+
+        ResponseEntity finalResponse = caseDocumentClientApi.uploadDocuments(
+            files,
+            CLASSIFICATION,
+            CASE_TYPE_ID,
+            BEARER + TOKEN
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:96]: Can be single constant BEARER + TOKEN

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+        files.add(getMultiPartFile("testFile2"));
+        files.add(getMultiPartFile("testFile3"));
+
+        ResponseEntity response = new ResponseEntity(HttpStatus.OK);
+
+        stubForUpload(response);
+
+        ResponseEntity finalResponse = caseDocumentClientApi.uploadDocuments(
+            files,
+            CLASSIFICATION,
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:116]: Please add matchers in the wiremock for all these params.

**Review by smathangi — COMMENTED** [review]
few comments @KleoG

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+            CASE_TYPE_ID,
+            BEARER + TOKEN
+        );
+
+        assertEquals(HttpStatus.OK, finalResponse.getStatusCode());
+        assertEquals(1, files.size());
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:100]: files assert checking nothing

Comment (KleoG) [reply src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:100]: yep was supposed to remove this before committing will remove

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+    @Test
+    void uploadMultipleDocumentTest() throws IOException {
+        List<MultipartFile> files = new ArrayList<>();
+        files.add(getMultiPartFile("testFile1"));
+        files.add(getMultiPartFile("testFile2"));
+        files.add(getMultiPartFile("testFile3"));
+
+        ResponseEntity response = new ResponseEntity(HttpStatus.OK);
+
+        stubForUpload(response);
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:112]: ideally pass all request params to stubForUpload method and match all of them correctly. If your response is just plain http OK, then no need to pass here. Inside stubForUpload you can do just `    .willReturn(aResponse()
                                    .withStatus(HttpStatus.OK.value())`

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+                                    .withStatus(HttpStatus.OK.value())
+                                    .withHeader(CONTENT_TYPE, APPLICATION_JSON_VALUE)
+                                    .withBody(objectMapper.writeValueAsString(response))
+                    )
+        );
+    }
+
+
+
+    private MultipartFile getMultiPartFile(String name) {
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:267]: you can use existing `MockMultipartFile` instead

File: src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java (hunk 3 above)
Comment (KleoG) [reply src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java:9]: Need no args constructor for one of the tests

Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/DocumentTTLResponse.java:9]: use this https://github.com/hmcts/idam-java-client/blob/master/src/test/java/uk/gov/hmcts/reform/idam/client/IdamClientTest.java#L88

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+            TOKEN,
+            SERVICE_AUTHORISATION_VALUE,
+            DOCUMENT_ID
+        );
+
+        assertEquals(HttpStatus.OK, finalResponse.getStatusCode());
+    }
+
+    @Test
+    void getDocumentMetaData() throws IOException {
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:113]: could you name test properly.

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+            TOKEN,
+            SERVICE_AUTHORISATION_VALUE,
+            DOCUMENT_ID
+        );
+
+        assertEquals(finalResponse.createdOn, response.createdOn);
+    }
+
+    @Test
+    void shouldDeleteDocument() {
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:131]: same as above ^^

File: src/test/resources/application.yml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+---
+case_document_am:
+    url: 'http://localhost:5050'
```
Comment (smathangi) [inline src/test/resources/application.yml:3]:
```suggestion
    url: 'http://localhost:${wiremock.server.port:5050}'
```

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+        );
+
+        assertEquals(HttpStatus.OK, finalResponse.getStatusCode());
+    }
+
+    @Test
+    void shouldSuccessfullyGetMetadataForDocument() throws IOException {
+        Date createdOn = Date.from(Instant.now());
+
+        Document response = new Document();
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:118]: please set all other fields and assert them. Define a builder in Document if it's not there..

File: src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+            urlPathEqualTo(URL + "/" + DOCUMENT_ID))
+               .withQueryParam(PERMANENT_QUERY_PARAM, equalTo(String.valueOf(PERMANENT))));
+    }
+
+    @Test
+    void shouldSuccessfullyPatchDocument() throws IOException {
+        LocalDateTime localDateTime = LocalDateTime.now();
+
+        DocumentTTLRequest request = new DocumentTTLRequest(localDateTime);
+        DocumentTTLResponse response = new DocumentTTLResponse(null, null, localDateTime);
```
Comment (smathangi) [inline src/test/java/uk/gov/hmcts/reform/ccd/document/am/feign/CaseDocumentClientApiTest.java:154]: set other fields some value and assert them

**Review by smathangi — APPROVED** [review]

**Review by kiran-yenigala-hmcts — APPROVED** [review]
LGTM

**Pull request merged**